File: examples/java/logging.java
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */



/**
 * @file logging.java
 *
 * This file contains an example outlining the use of the GMSEC API logging
 * framework.
 */

import gov.nasa.gsfc.gmsec.api.*;
import gov.nasa.gsfc.gmsec.api.util.Log;
import gov.nasa.gsfc.gmsec.api.util.LogLevel;
import gov.nasa.gsfc.gmsec.api.util.LogHandler;
import gov.nasa.gsfc.gmsec.api.util.LogEntry;
import gov.nasa.gsfc.gmsec.api.util.TimeUtil;
import gov.nasa.gsfc.gmsec.api.mist.ConnectionManager;
import gov.nasa.gsfc.gmsec.api.mist.ConnectionManagerCallback;
import java.util.concurrent.Semaphore;

//o Define a baseline LogHandler
// This will be used by the Log macros; The implementation of onMessage
// determines how messages will be logged to output
class BaseHandler extends LogHandler
{
	// Use a mutex so messages from different threads don't get mixed up
	private Semaphore mutex;
	private String whoAmI = "";

	public BaseHandler(String wai)
	{
		mutex = new Semaphore(1);
		whoAmI = wai;
	}

	public void onMessage(LogEntry entry)
	{
		try
		{
			mutex.acquire();

			String time = TimeUtil.formatTime(entry.time);

			System.out.println("[BaseHandler::onMessage] for: " + whoAmI +
				 " : " + time + " [" + Log.levelToString(entry.level) + "]" +
				 " [" + entry.fileName + ":" + entry.lineNumber + "] " +
				 entry.message + "\n\n");

			mutex.release();
		}
		catch(Exception e)
		{
			Log.error(e.getMessage());
		}
	}
}

class ErrorHandler extends BaseHandler
{
	public ErrorHandler()
	{
		super("GMSEC_ERROR");
	}
}

class WarningHandler extends BaseHandler
{
	public WarningHandler()
	{
		super("GMSEC_WARNING");
	}
}

class InfoHandler extends BaseHandler
{
	public InfoHandler()
	{
		super("GMSEC_INFO");
	}
}

class VerboseHandler extends BaseHandler
{
	public VerboseHandler()
	{
		super("GMSEC_VERBOSE");
	}
}

class DebugHandler extends BaseHandler
{
	public DebugHandler()
	{
		super("GMSEC_DEBUG");
	}
}

class AnyHandler extends BaseHandler
{
	public AnyHandler()
	{
		super("ANY_HANDLER");
	}
}

public class logging
{
	public static void main(String[] args)
	{
		if (args.length < 1)
		{
			System.out.println("usage: java logging mw-id=<middleware ID>");
			System.exit(-1);
		}

		Config config = new Config(args);

		//o Create and register log handlers
		ErrorHandler errorHandler = new ErrorHandler();
		WarningHandler warningHandler = new WarningHandler();
		InfoHandler infoHandler = new InfoHandler();
		VerboseHandler verboseHandler = new VerboseHandler();
		DebugHandler debugHandler = new DebugHandler();
		AnyHandler anyHandler = new AnyHandler();

		Log.registerHandler(LogLevel.ERROR, errorHandler);
		Log.registerHandler(LogLevel.WARNING, warningHandler);
		Log.registerHandler(LogLevel.INFO, infoHandler);
		Log.registerHandler(LogLevel.VERBOSE, verboseHandler);
		Log.registerHandler(LogLevel.DEBUG, debugHandler);

		//o Set logging reporting level
		Log.setReportingLevel(LogLevel.VERBOSE);
		Log.verbose("The log reporting level is now set to: " + Log.getReportingLevel());

		//o Print the GMSEC API version number using the GMSEC Logging
		// interface
		// This is useful for determining which version of the API is
		// configured within the environment
		Log.info(ConnectionManager.getAPIVersion());

		try
		{
			//o Create the ConnectionManager
			ConnectionManager connMgr = new ConnectionManager(config);

			//o Connect
			connMgr.initialize();

			//o Output middleware client library version
			Log.info(connMgr.getLibraryVersion());

			//o Publish a message
			publishTestMessage(connMgr, "GMSEC.TEST.PUBLISH");

			//o Disconnect from the middleware and clean up the Connection
			connMgr.cleanup();
		}
		catch (Exception e)
		{
			Log.error(e.getMessage());
			System.exit(-1);
		}

		//o Unregister log handlers
		Log.registerHandler(null);

		//o Set log stream to stderr
		config.addValue("LOGFILE", "STDERR");
		Log.info("This message should go to stderr, not stdout.  " +
						 "For example, in bash test by running as:\n" +
						 "java logging mw-id=bolt 2> testfile.txt\n" +
						 "... and then check the contents of testfile.txt");

		//o Reset log stream to stdout
		config.addValue("LOGFILE", "STDOUT");

		Log.error("This is an example error message.");
		Log.warning("This is an example warning message.");
		Log.verbose("This is an example \"verbose\" message.");
		Log.debug("This is an example debug message which should not show.");

			// This last message cannot be shown right now because
	  	// Log::setReportingLevel(logVERBOSE), used above, does not
	  	// allow DEBUG messages to come out.
		Log.verbose("This is another example \"verbose\" message.");

		// Set logging reporting level to now allow DEBUG messages to be shown
		Log.setReportingLevel(LogLevel.DEBUG);
		if(Log.getReportingLevel() == LogLevel.DEBUG)
		{
			Log.info("Changed reporting level to logDEBUG");
		}
		else
		{
			Log.error("Failed to change reporting level to logDEBUG");
		}
		// The DEBUG message below will be shown successfully, unlike the last
		// debug message.
		Log.debug( "This is an example debug message which should show.");

		Log.debug("NONE reporting level, numerically, is " + Log.levelFromString("NONE"));
		Log.debug("ERROR reporting level, numerically, is " + Log.levelFromString("ERROR"));
		Log.debug("SECURE reporting level, numerically, is " + Log.levelFromString("SECURE"));
		Log.debug("WARNING reporting level, numerically, is " + Log.levelFromString("WARNING"));
		Log.debug("INFO reporting level, numerically, is " + Log.levelFromString("INFO"));
		Log.debug("VERBOSE reporting level, numerically, is " + Log.levelFromString("VERBOSE"));
		Log.debug("DEBUG reporting level, numerically, is " + Log.levelFromString("DEBUG"));

		//o Register general-purpose handler and test
		Log.registerHandler(anyHandler);

		Log.error("NONE reporting level, numerically, is " + Log.levelFromString("NONE"));
		Log.error("ERROR reporting level, numerically, is " + Log.levelFromString("ERROR"));
		Log.warning("WARNING reporting level, numerically, is " + Log.levelFromString("WARNING"));
		Log.info("INFO reporting level, numerically, is " + Log.levelFromString("INFO"));
		Log.verbose("VERBOSE reporting level, numerically, is " + Log.levelFromString("VERBOSE"));
		Log.debug("DEBUG reporting level, numerically, is " + Log.levelFromString("DEBUG"));

		//o Unregister log handlers
		Log.registerHandler(null);

		return;
	}

	public static void publishTestMessage(ConnectionManager connMgr, String subject)
	{
		int i = 123;
		try
		{
			//o Create a Message object
			Message message = new Message(subject, Message.MessageKind.PUBLISH);

			//o Add fields to the Message
			message.addField("F", false);
			message.addField("I", (int) i); //I32
			message.addField("K", (short) i); //I16
			message.addField("S", "This is a test");
			message.addField("D", (float) (1 + 1./i)); //F32
			message.addField("X", "JLMNOPQ".getBytes()); //Bin

			//o Publish Message
			connMgr.publish(message);

			//o Output the Message's XML string representation by invoking Log macro
			Log.info("Sent:\n" + message.toXML());
		}
		catch(Exception e)
		{
			Log.error(e.getMessage());
		}
	}
}

File: examples/python3/logging.py
#!/usr/bin/env python3

# Copyright 2007-2021 United States Government as represented by the
# Administrator of The National Aeronautics and Space Administration.
# No copyright is claimed in the United States under Title 17, U.S. Code.
# All Rights Reserved.


""" 
 @file logging.py
 
 This file contains an example outlining the use of the GMSEC API logging
 framework.
"""

import libgmsec_python3
import sys

# Define a baseline LogHandler
# This will be used by the Log macros; The implementation of on_message
# determines how messages will be logged to output

class BaseHandler(libgmsec_python3.LogHandler):

    def __init__(self, wai):
        # Use a mutex so messages from different threads don't get mixed up
        libgmsec_python3.LogHandler.__init__(self)
        self.mutex = libgmsec_python3.Mutex()
        self.whoami = wai

    def on_message(self, entry):
        self.mutex.enter_mutex()

        tempBuffer = libgmsec_python3.TimeUtil.format_time(entry.time)

        msg = "[BaseHandler::on_message] for: " + self.whoami \
            + " : " + tempBuffer + " [" + libgmsec_python3.Log.to_string(entry.level) + "]" \
            + " [" + entry.file + ":" + str(entry.line) + "] " + entry.message + "\n" 

        print(msg.replace("\n", "\n\t"))

        self.mutex.leave_mutex()


# Different Handlers can be implemented for each logging level, if so desired
class ErrorHandler(BaseHandler):
    def __init__(self):
        BaseHandler.__init__(self,"GMSEC_ERROR")

class WarningHandler(BaseHandler):
    def __init__(self):
        BaseHandler.__init__(self,"GMSEC_WARNING")

class InfoHandler(BaseHandler):
    def __init__(self):
        BaseHandler.__init__(self,"GMSEC_INFO")

class AnyHandler(BaseHandler):
    def __init__(self):
        BaseHandler.__init__(self,"ANY_HANDLER") 


def main(agrv=None):

    if(len(sys.argv) <= 1):
        usageMessage =  "usage: " + sys.argv[0] + " mw-id=<middleware ID>"
        print(usageMessage)
        return -1

    config = libgmsec_python3.Config(sys.argv)

    # Create and register log handlers
    errorHandler   = ErrorHandler()
    warningHandler = WarningHandler()
    infoHandler    = InfoHandler()
    anyHandler     = AnyHandler()

    libgmsec_python3.Log.register_handler(errorHandler,   libgmsec_python3.logERROR)
    libgmsec_python3.Log.register_handler(warningHandler, libgmsec_python3.logWARNING)
    libgmsec_python3.Log.register_handler(infoHandler,    libgmsec_python3.logINFO)
    # Note: At this time, for the Python3 binding, setting log handlers for logVERBOSE and logDEBUG are not permitted,
    # This is because certain middleware wrappers output log messages after the custom Python log handler
    # has been garbage collected, and this can lead to fatal errors.
    libgmsec_python3.Log.register_handler(anyHandler, libgmsec_python3.logVERBOSE)   # has no effect, but shown for demo purposes
    libgmsec_python3.Log.register_handler(anyHandler,   libgmsec_python3.logDEBUG)     # has no effect, but shown for demo purposes

    # Set logging reporting level
    libgmsec_python3.Log.set_reporting_level(libgmsec_python3.logVERBOSE)
    libgmsec_python3.log_verbose("The log reporting level is now set to: " + str(libgmsec_python3.Log.get_reporting_level()))

    # Print the GMSEC API version number using the GMSEC Logging
    # interface
    # This is useful for determining which version of the API is
    # configured within the environment
    libgmsec_python3.log_info(libgmsec_python3.ConnectionManager.get_API_version())

    try:
        # Create the ConnectionManager
        connMgr = libgmsec_python3.ConnectionManager(config)

        # Connect
        connMgr.initialize()

        # Output middleware client library version
        libgmsec_python3.log_info(connMgr.get_library_version())

        # Publish a message
        publishTestMessage(connMgr, "GMSEC.TEST.PUBLISH")

        # Disconnect from the middleware and clean up the Connection
        connMgr.cleanup()
        
    except libgmsec_python3.GmsecError as e:
        libgmsec_python3.log_error(str(e))
        return -1
        

    # Unregister log handlers
    libgmsec_python3.Log.register_handler(None)

    # Set log stream to stderr
    config.add_value("LOGFILE", "STDERR")
    libgmsec_python3.log_info("This message should go to stderr, not stdout.  " + "For example, in bash test by running as:\n" + "./logging mw-id=bolt 2> testfile.txt\n" + "... and then check the contents of testfile.txt")

    # Reset log stream to stdout
    config.add_value("LOGFILE", "STDOUT")

    libgmsec_python3.log_error("This is an example error message.")
    libgmsec_python3.log_warning("This is an example warning message.")
    libgmsec_python3.log_verbose("This is an example \"verbose\" message.")
    libgmsec_python3.log_debug("This is an example debug message which should not show.")

    # This last message cannot be shown right now because
    # Log::set_reporting_level(logVERBOSE), used above, does not
    # allow DEBUG messages to come out.
    libgmsec_python3.log_verbose("This is another example \"verbose\" message.")

    # Set logging reporting level to now allow DEBUG messages to be shown
    libgmsec_python3.Log.set_reporting_level(libgmsec_python3.logDEBUG)
    if (libgmsec_python3.Log.get_reporting_level() == libgmsec_python3.logDEBUG):
        libgmsec_python3.log_info("Changed reporting level to logDEBUG")
        
    else:
        libgmsec_python3.log_error("Failed to change reporting level to logDEBUG")
        
    # The DEBUG message below will be shown successfully, unlike the last
    # debug message.
    libgmsec_python3.log_debug("This is an example debug message which should show.")

    libgmsec_python3.log_debug("NONE reporting level, numerically, is " + str(libgmsec_python3.Log.from_string("NONE")))
    libgmsec_python3.log_debug("ERROR reporting level, numerically, is " + str(libgmsec_python3.Log.from_string("ERROR")))
    libgmsec_python3.log_debug("SECURE reporting level, numerically, is "  + str(libgmsec_python3.Log.from_string("SECURE")))
    libgmsec_python3.log_debug("WARNING reporting level, numerically, is " + str(libgmsec_python3.Log.from_string("WARNING")))
    libgmsec_python3.log_debug("INFO reporting level, numerically, is " + str(libgmsec_python3.Log.from_string("INFO")))
    libgmsec_python3.log_debug("VERBOSE reporting level, numerically, is " + str(libgmsec_python3.Log.from_string("VERBOSE")))
    libgmsec_python3.log_debug("DEBUG reporting level, numerically, is " + str(libgmsec_python3.Log.from_string("DEBUG")))

    # Register general-purpose handler and test
    try:
        libgmsec_python3.Log.register_handler(anyHandler)

    except libgmsec_python3.GmsecError as e:
        print(str(e))

    libgmsec_python3.log_error("NONE reporting level, numerically, is " + str(libgmsec_python3.Log.from_string("NONE")))
    libgmsec_python3.log_error("ERROR reporting level, numerically, is " + str(libgmsec_python3.Log.from_string("ERROR")))
    libgmsec_python3.log_warning("WARNING reporting level, numerically, is " + str(libgmsec_python3.Log.from_string("WARNING")))
    libgmsec_python3.log_info("INFO reporting level, numerically, is " + str(libgmsec_python3.Log.from_string("INFO")))
    libgmsec_python3.log_verbose("VERBOSE reporting level, numerically, is " + str(libgmsec_python3.Log.from_string("VERBOSE")))
    libgmsec_python3.log_debug("DEBUG reporting level, numerically, is " + str(libgmsec_python3.Log.from_string("DEBUG")))

    # Unregister log handlers
    libgmsec_python3.Log.register_handler(None)

    return 0


def publishTestMessage(connMgr, subject):

    i = 123

    # Create a Message object
    message = libgmsec_python3.Message(subject, libgmsec_python3.Message.PUBLISH)

    # Add fields to the Message
    message.add_field("F", False)
    message.add_field(libgmsec_python3.I32Field("I", i))
    message.add_field(libgmsec_python3.U16Field("K", i))
    message.add_field("S", "This is a test")
    message.add_field(libgmsec_python3.F32Field("D", 1 + 1./i))
    message.add_field(libgmsec_python3.BinaryField("X", "JLMNOPQ", 7))

    # Publish Message 
    connMgr.publish(message)

    # Output the Message's XML string representation by invoking Log macro
    libgmsec_python3.log_info("Sent:\n" + message.to_XML())


#
# Main entry point of script
#
if __name__=="__main__":
    sys.exit(main())

